### PR TITLE
Taxonomy: "Taxons with" facet should not return data for intermediate…

### PR DESCRIPTION
**Pull request by LeonardoGonzales**
Taxonomy: "Taxons with" facet should not return data for intermediate lineage nodes (TRM-27102)

### CODE REVIEW

File: spark-indexer/src/main/java/org/uniprot/store/spark/indexer/taxonomy/mapper/TaxonomyStatisticsAggregationMapper.java [hunk 1]
```diff
@@ -4,20 +4,25 @@
...
-        implements Function2<TaxonomyStatistics, TaxonomyStatistics, TaxonomyStatistics> {
+        implements Function2<
+                TaxonomyStatisticsWrapper, TaxonomyStatisticsWrapper, TaxonomyStatisticsWrapper> {
 
     private static final long serialVersionUID = -2085055845000585882L;
 
     @Override
-    public TaxonomyStatistics call(TaxonomyStatistics stat1, TaxonomyStatistics stat2)
+    public TaxonomyStatisticsWrapper call(
+            TaxonomyStatisticsWrapper wrapper1, TaxonomyStatisticsWrapper wrapper2)
```
Comment (eddturner) [inline spark-indexer/src/main/java/org/uniprot/store/spark/indexer/taxonomy/mapper/TaxonomyStatisticsAggregationMapper.java:17]: are wrapper1 and 2 relating to the organism + proteome tax wrappers? If so, could we name these variables differently, to reflect this?

File: spark-indexer/src/main/java/org/uniprot/store/spark/indexer/taxonomy/mapper/TaxonomyStatisticsAggregationMapper.java [hunk 2]
```diff
@@ -4,20 +4,25 @@
...
             throws Exception {
-        TaxonomyStatistics mergedStat = null;
-        if (SparkUtils.isThereAnyNullEntry(stat1, stat2)) {
-            mergedStat = SparkUtils.getNotNullEntry(stat1, stat2);
+        TaxonomyStatisticsWrapper mergedWrapper = null;
+        if (SparkUtils.isThereAnyNullEntry(wrapper1, wrapper2)) {
+            mergedWrapper = SparkUtils.getNotNullEntry(wrapper1, wrapper2);
         } else {
-            mergedStat =
+            TaxonomyStatistics stat1 = wrapper1.getStatistics();
```
Comment (eddturner) [inline spark-indexer/src/main/java/org/uniprot/store/spark/indexer/taxonomy/mapper/TaxonomyStatisticsAggregationMapper.java:23]: can we name stat1/2 differently to reflect the statistics they're for?

File: spark-indexer/src/main/java/org/uniprot/store/spark/indexer/taxonomy/mapper/TaxonomyStatisticsWrapperJoinMapper.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+public class TaxonomyStatisticsWrapperJoinMapper
+        implements Function<
+                Tuple2<Optional<TaxonomyStatisticsWrapper>, Optional<TaxonomyStatisticsWrapper>>,
+                TaxonomyStatisticsWrapper> {
+
+    private static final long serialVersionUID = 7240080849019578656L;
+
+    @Override
+    public TaxonomyStatisticsWrapper call(
```
Comment (eddturner) [inline spark-indexer/src/main/java/org/uniprot/store/spark/indexer/taxonomy/mapper/TaxonomyStatisticsWrapperJoinMapper.java:19]: after thinking about the class name, I think *Wrapper is fine. It's a wrapper, after all. We could call it something like *Info, but arguably, it's not clear that it wraps a TaxStats object... therefore, I think TaxonomyStatisticsWrapper is good.

**Review by eddturner — APPROVED** [review]
Approved, but with a couple of optional suggestions. Really nice and yet complicated work you did here!

File: spark-indexer/src/main/java/org/uniprot/store/spark/indexer/taxonomy/mapper/TaxonomyStatisticsAggregationMapper.java (hunk 2 above)
Comment (LeonardoGonzales) [reply spark-indexer/src/main/java/org/uniprot/store/spark/indexer/taxonomy/mapper/TaxonomyStatisticsAggregationMapper.java:23]: Hi Edd, This Statistic Aggregation Mapper is used to aggregate, protein stats with another protein stats, and also is used to aggregate proteome stats with another proteome stats. This is why it is difficult to get a good name for it.

Comment (LeonardoGonzales) [reply spark-indexer/src/main/java/org/uniprot/store/spark/indexer/taxonomy/mapper/TaxonomyStatisticsAggregationMapper.java:23]: I will merge it and we can try to figure out a better name for it afterwards.

**Pull request merged**